explainer.md

```diff
@@ -22,118 +71,260 @@ place associated with the application (such as the shelf or home screen).
 
 The purpose of this API is:
 
-* To give the application a small but visible place to subtly notify the user
-  that there is new activity that might require their attention, without showing
-  a full [notification](https://notifications.spec.whatwg.org/).
+* To give web sites a small but visible place to subtly notify the user that there is new activity that might require their attention without showing a full [notification](https://notifications.spec.whatwg.org/).
 * To indicate a small amount of additional information, such as an unread count.
-* To allow the application to convey this information regardless of whether any
-  of the application's windows are open.
+* To allow certain user blessed pages (such as Bookmarks or [Installed Web Applications](https://www.w3.org/TR/appmanifest/#installable-web-applications)) to convey this information, regardless of whether they are currently open.
 
 Non-goals are:
 
 * To provide an arbitrary image badge.
 
 Possible areas for expansion:
 
-* Providing badging for sites in a normal web browsing context. The current
-  proposal is just for installed apps (designed to show up in the operating
-  system shelf area). We could also explore icon badging on the drive-by web.
-  This naturally leads into...
-* Providing per-tab and per-window badging. The current proposal is for a global badge
-  for the application. See [#1](https://github.com/WICG/badging/issues/1).
-* Support apps that want to render a small status indicator (e.g., a music app shows ▶️	
-  or ⏸️; a weather app shows ⛈️ or ⛅️).
-* Setting the badge from a service worker (e.g. an email app updating an unread count).
+* Support rendering a small status indicator (e.g., a music app shows ▶️ or ⏸️; a weather app shows ⛈️ or ⛅️).
+* Support setting the badge from a service worker (e.g. an email app updating an unread count in the background).
+* Support badges on bookmark icons (this is probably up to user agents to support, and may not need anything new API wise).
+* Support for non-inherited badges (see [Issue 42](https://github.com/WICG/badging/issues/42)).
+* Support for unicode glyph badges.
 
 Examples of sites that may use this API:
 
-* Chat, email and social apps, to signal that new messages have arrived.
-* Productivity apps, to signal that a long-running background task (such as
-  rendering an image or video) has completed.
-* Games, to signal that a player action is required (e.g., in Chess, when it is
-  the player's turn).
+* Chat, email and social apps; to signal that new messages have arrived.
+* Productivity apps, to signal that a long-running background task (such as rendering an image or video) has completed.
```

Comment:
This one doesn't fit, tbh... at least, with the current numbering scheme.

Reply:
I think @mgiuca meant that you could call `Badge.set()` when the task completes. It seems like a better fit for a notification though, TBH. Reckon I should remove it?

Reply:
I guess... that's not traditionally how badges are used (as they generally imply a "call to action" by the user). That is, I've never seen a badge count down in the background. But I guess if they can count up (e.g., I get more emails), then can also count down to zero.

Reply:
It wasn't meant to be a countdown showing on the badge. More of a little dot to indicate "your coffee is ready".

I agree a notification would be more appropriate for something like this, but maybe you want a low-priority notification rather than one that always pops up. You could think of it as a "call to action" because if you were waiting for a task to complete, you can't use the app, but now you can get back to work.

Reply:
hmmm... maybe that's what `set(0)` could be instead of clear? Just a dot, no number.

Reply:
That's what `set()` is for :)

Reply:
We were debating exactly how you would set a dot (maybe `set(true)` or even `setFlag()` was proposed by Domenic), but we want `set(0)` to be the same thing as clearing, so that you can call `set(unreadCount)` without having to write `if (unreadCount > 0) set(unreadCount); else clear();`.

(Assuming we don't actually want to allow "0" to be displayed.) I wouldn't want `set(0)` to do something undesirable in the case where you had no unread messages.

Reply:
I'm still a bit confused why we need `.clear()` if `.set(0)` does the same thing? With respect to `setFlag()`, it seems to go counter having functions that do entirely different things.    

About: 

```JS
if (unreadCount > 0) set(unreadCount); else clear();
```

It seems quite application specific. What about what @fallaciousreasoning suggested?:

```JS
Badge.set(); // show indicator, no number.
Badge.set(0); // clear.
Badge.set(123); // show number
```

Reply:
^ Note, this isn't my suggestion, this is the current API proposal (as @mgiuca originally wrote it), though there is also an explicit `Badge.clear()` which feels fairly natural.

Reply:
I think `Badge.clear()` feels like a fairly natural counterpart to `Badge.set()`, though I'm not opposed to
```js
Badge.set(true) // Sets a flag.
Badge.set(false) // Clears flag
```

as `false` is to the `boolean` space as `0` is to the `integer` space.

Interestingly, I suspect `Badge.set(false)` will already clear the badge (because it will be cast to the number 0). However `Badge.set(true)` will set the badge to `1`, not flag.

Perhaps we could completely avoid the whole flag/integer debate if we only supported numerical badges (the value `1` could just be displayed as a flag, if the user-agent so chose). This way `true`/`false` would Just Work™ and we would display integers for n > 1.

Reply:
> It seems quite application specific.

@marcoscaceres The intended usage is in the current explainer under "API Proposal".

The idea was that there would be two primary usage "modes" for this API: integer "mode" and Boolean "mode" (plus other symbols but let's ignore that for the time being). In Integer "mode" you always pass an int, which reflects an unread count or similar:

```js
Badge.set(getUnreadCount());
```

In this mode of operation, the badge is implicitly cleared if the unread count is 0. I think that's pretty reasonable behaviour. I could be convinced that we should just explicitly show "0" if you pass `0`, but I think that would be more surprising than desirable.

In Boolean "mode" you always either call `set()` with no parameters, or `clear()`:

```js
if (myTurn())
  Badge.set();
else
  Badge.clear();
```

The reason we have a `clear()` is because, even though it does the same thing as `set(0)`, it looks super weird to write "`if (condition) Badge.set(); else Badge.set(0)`".

We could also remove `clear` and just have `set(true)` to set the flag and `set(false)` to clear it, but I don't like the idea that `set(true)` and `set(1)` give different behaviour. That's why we didn't originally specify the Boolean mode as being literally passing a Boolean.

> Interestingly, I suspect `Badge.set(false)` will already clear the badge (because it will be cast to the number 0). However `Badge.set(true)` will set the badge to 1, not flag.

Yep, that's what will happen, but only because we defined the API as taking a `long`. We _could_ theoretically make it "`long or boolean`" and then `true` and `1` would be treated as independent values. But as I said above, I think that's confusing.

> Perhaps we could completely avoid the whole flag/integer debate if we only supported numerical badges (the value `1` could just be displayed as a flag, if the user-agent so chose). This way `true`/`false` would Just Work™ and we would display integers for n > 1.

Hmm, interesting proposal, but I think if the app wants to display a number, we should allow it to display "1" when there's one unread mail. And if an app wants to display a Boolean, we shouldn't allow the UA to show "1" or empty. I think we want explicit separate integer and Boolean badge types.

Reply:
> Hmm, interesting proposal, but I think if the app wants to display a number, we should allow it to display "1" when there's one unread mail. And if an app wants to display a Boolean, we shouldn't allow the UA to show "1" or empty. I think we want explicit separate integer and Boolean badge types.

Suits me :)

Reply:
> I think Badge.clear() feels like a natural counterpart to Badge.set(), 

I do like .clear() as it's obvious what it does.  

> We could also remove clear and just have set(true) to set the flag and set(false) to clear it, but I don't like the idea that set(true) and set(1) give different behaviour. ...  But as I said above, I think that's confusing.

Agree, would be confusing and would violate a JS invariant: true/false always coerce to 1/0.

> we should allow it to display "1" when there's one unread mail

Agree. Definitely want to keep the ability to display "1". 

I'm liking the modes thing... so, I think I'm sold on what we currently have 👍

Reply:
Awesome, I'm resolving this :)